### PR TITLE
Mes 2313 send tests in interval

### PR DESCRIPTION
**Pull request by szabolcs-keresztes**
### Refactor test statuses

- Making so that the status of a test can be changed using the slotId to identify the test
- Extracting some of the status changes to the effects
- Also setting the Decided status of a test when we End Debrief

### Send tests in interval

- Setting up an interval which dispatches a send completed tests action every 15 mins (this is configured in the configuration microservice)
- After the test have been sent we change the status of those tests to Submitted

### CODE REVIEW

File: src/modules/tests/test-status/test-status.actions.ts [hunk 1]
```diff
@@ -1,30 +1,39 @@
 import { Action } from '@ngrx/store';
 
-// TODO: These actions are not relatable to an action in the app. Rename needed.
+export const SET_TEST_STATUS_BOOKED = '[???] Set Test Status to Booked';
+export const SET_TEST_STATUS_STARTED = '[???] Set Test Status to Started';
+export const SET_TEST_STATUS_DECIDED = '[???] Set Test Status to Decided';
+export const SET_TEST_STATUS_COMPLETED = '[???] Set Test Status to Completed';
+export const SET_TEST_STATUS_SUBMITTED = '[???] Set Test Status to Submitted';
```
Comment (rossfauldsbjss) [inline src/modules/tests/test-status/test-status.actions.ts]: [???] needs proper name

Comment (rossfauldsbjss) [reply src/modules/tests/test-status/test-status.actions.ts]: And I see the TODO above - so ignore me.

Comment (rossfauldsbjss) [inline src/modules/tests/test-status/test-status.actions.ts]: [???] Todo needs addressing.

File: src/modules/tests/test-status/test-status.actions.ts [hunk 2]
```diff
@@ -1,30 +1,39 @@
 import { Action } from '@ngrx/store';
 
-// TODO: These actions are not relatable to an action in the app. Rename needed.
+export const SET_TEST_STATUS_BOOKED = '[???] Set Test Status to Booked';
```
Comment (Cavalier07) [inline src/modules/tests/test-status/test-status.actions.ts]: Need a name for these - can't it just be [ Test Status ]

File: src/modules/tests/test-status/test-status.reducer.ts [hunk 3]
```diff
@@ -2,24 +2,38 @@ import { TestStatus } from './test-status.model';
...
+        ...state,
+        [action.slotId]: TestStatus.Submitted,
+      };
     default:
       return state;
   }
 }
 
-export const getCandidate = createFeatureSelector<TestStatus>('testStatus');
+export const getTestStatus = createFeatureSelector<TestStatus>('testStatus');
```
Comment (Cavalier07) [inline src/modules/tests/test-status/test-status.reducer.ts]: Nice Spot :)

File: src/modules/tests/tests.actions.ts [hunk 4]
```diff
@@ -3,6 +3,10 @@ import { TestsModel } from './tests.model';
 import { ActivityCode, StandardCarTestCATBSchema } from '@dvsa/mes-test-schema/categories/B';
 
 export const START_SENDING_COMPLETED_TESTS = '[???] Start Sending Completed Test';
```
Comment (Cavalier07) [inline src/modules/tests/tests.actions.ts]: Can we fix the ??? here

File: src/modules/tests/tests.selector.ts [hunk 5]
```diff
@@ -14,14 +14,18 @@ const outcomeStatus: any[] = [
   { outcomeCode: '5', outcomeText: 'Unsuccessful', outcomeColour: 'mes-red' },
 ];
 
+export const getCurrentTestSlotId = (tests: TestsModel) => {
+  return tests.currentTest.slotId;
+};
+
```
Comment (Cavalier07) [inline src/modules/tests/tests.selector.ts]: can we shorten this?

`export const getCurrentTestSlotId = (tests: TestsModel) => tests.currentTest.slotId;`

File: src/modules/tests/test-status/test-status.actions.ts (hunk 1 above)
Comment (szabolcs-keresztes) [reply src/modules/tests/test-status/test-status.actions.ts]: ahh thanks!

File: src/pages/debrief/debrief.ts [hunk 6]
```diff
@@ -199,7 +198,7 @@ export class DebriefPage extends BasePageComponent {
       this.navController.popToRoot();
       return;
     }
-    this.store$.dispatch(new PersistTests());
+    this.store$.dispatch(new EndDebrief());
```
Comment (Cavalier07) [inline src/pages/debrief/debrief.ts]: Much Better way of doing it :)

File: src/pages/login/login.ts [hunk 7]
```diff
@@ -94,6 +94,9 @@ export class LoginPage extends BasePageComponent {
 
       this.analytics.initialiseAnalytics();
       this.store$.dispatch(new StartSendingLogs());
+
+      this.store$.dispatch(new StartSendingCompletedTests());
```
Comment (Cavalier07) [inline src/pages/login/login.ts]: Missing Tests for this - Ammar added a load as this kept causing issues

File: src/pages/debrief/debrief.effects.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import { Injectable } from '@angular/core';
+import { Effect, Actions, ofType } from '@ngrx/effects';
```
Comment (Cavalier07) [inline src/pages/debrief/debrief.effects.ts]: Need Tests for this

File: src/pages/office/office.effects.ts [hunk 9]
```diff
@@ -29,6 +36,23 @@ export class OfficeEffects {
         testSummaryActions.D255_YES,
         testSummaryActions.D255_NO,
         ),
-    map(() => new PersistTests()),
+    map(() => new testsActions.PersistTests()),
+  );
+
+  @Effect()
+  completeTestEffect$ = this.actions$.pipe(
```
Comment (Cavalier07) [inline src/pages/office/office.effects.ts]: Need Tests for this

File: src/pages/waiting-room/waiting-room.effects.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import { Injectable } from '@angular/core';
+import { Effect, Actions, ofType } from '@ngrx/effects';
```
Comment (Cavalier07) [inline src/pages/waiting-room/waiting-room.effects.ts]: Need Tests

File: src/modules/tests/test-status/test-status.actions.ts (hunk 1 above)
Comment (szabolcs-keresztes) [reply src/modules/tests/test-status/test-status.actions.ts]: Added the names of the components that dispatch these actions

File: src/modules/tests/test-status/test-status.actions.ts (hunk 2 above)
Comment (szabolcs-keresztes) [reply src/modules/tests/test-status/test-status.actions.ts]: done

File: src/modules/tests/tests.actions.ts (hunk 4 above)
Comment (szabolcs-keresztes) [reply src/modules/tests/tests.actions.ts]: Sure added a good name

**Pull request merged**